rwkv.cpp

```diff
@@ -265,194 +465,382 @@ struct ggml_tensor * rwkv_layer_norm(ggml_context * ctx, struct ggml_tensor * x,
+}
+
+void ctx_size_add_tensor(struct ctx_size & size, const uint64_t tensors, const uint64_t views, const tensor_header & header) {
+    ctx_size_add_tensor(size, tensors, views, type_to_ggml[header.data_type], header.width, header.height);
+}
+
+struct ctx_size rwkv_single_att_size(const size_t n_embed = 0) {
+    size_t ptr_nelem = sizeof(void *) / sizeof(uint32_t);
+
+    struct ctx_size ctx_size;
+
+    /*  x0 */ ctx_size_add_tensor(ctx_size, 2, 1, GGML_TYPE_F32, n_embed);
+
+    /*  xk */ ctx_size_add_tensor(ctx_size, 3, 1, GGML_TYPE_F32, n_embed);
+    /*  xk */ ctx_size_add_tensor(ctx_size, 1, 0, GGML_TYPE_I32, ptr_nelem);
+    /*  xv */ ctx_size_add_tensor(ctx_size, 3, 1, GGML_TYPE_F32, n_embed);
+    /*  xv */ ctx_size_add_tensor(ctx_size, 1, 0, GGML_TYPE_I32, ptr_nelem);
+    /*  xr */ ctx_size_add_tensor(ctx_size, 3, 1, GGML_TYPE_F32, n_embed);
+    /*  xr */ ctx_size_add_tensor(ctx_size, 1, 0, GGML_TYPE_I32, ptr_nelem);
+
+    /*   r */ ctx_size_add_tensor(ctx_size, 2, 0, GGML_TYPE_F32, n_embed);
+    /*   r */ ctx_size_add_tensor(ctx_size, 1, 0, GGML_TYPE_I32, ptr_nelem);
+    /*   k */ ctx_size_add_tensor(ctx_size, 1, 0, GGML_TYPE_F32, n_embed);
+    /*   v */ ctx_size_add_tensor(ctx_size, 1, 0, GGML_TYPE_F32, n_embed);
+
+    /*  ww */ ctx_size_add_tensor(ctx_size, 1, 0, GGML_TYPE_F32, n_embed);
+    /*  qq */ ctx_size_add_tensor(ctx_size, 1, 0, GGML_TYPE_F32, n_embed);
+    /*  qq */ ctx_size_add_tensor(ctx_size, 1, 0, GGML_TYPE_I32, ptr_nelem);
+    /*  e1 */ ctx_size_add_tensor(ctx_size, 2, 0, GGML_TYPE_F32, n_embed);
+    /*  e1 */ ctx_size_add_tensor(ctx_size, 1, 0, GGML_TYPE_I32, ptr_nelem);
+    /*  e2 */ ctx_size_add_tensor(ctx_size, 2, 0, GGML_TYPE_F32, n_embed);
+    /*  e2 */ ctx_size_add_tensor(ctx_size, 1, 0, GGML_TYPE_I32, ptr_nelem);
+
+    /*   a */ ctx_size_add_tensor(ctx_size, 2, 1, GGML_TYPE_F32, n_embed);
+    /*   b */ ctx_size_add_tensor(ctx_size, 1, 1, GGML_TYPE_F32, n_embed);
+
+    /*  ww */ ctx_size_add_tensor(ctx_size, 1, 0, GGML_TYPE_F32, n_embed);
+    /*  qq */ ctx_size_add_tensor(ctx_size, 1, 0, GGML_TYPE_F32, n_embed);
+    /*  qq */ ctx_size_add_tensor(ctx_size, 1, 0, GGML_TYPE_I32, ptr_nelem);
+    /*  e1 */ ctx_size_add_tensor(ctx_size, 2, 0, GGML_TYPE_F32, n_embed);
+    /*  e1 */ ctx_size_add_tensor(ctx_size, 1, 0, GGML_TYPE_I32, ptr_nelem);
+    /*  e2 */ ctx_size_add_tensor(ctx_size, 2, 0, GGML_TYPE_F32, n_embed);
+    /*  e2 */ ctx_size_add_tensor(ctx_size, 1, 0, GGML_TYPE_I32, ptr_nelem);
+
+    /*  xx */ ctx_size_add_tensor(ctx_size, 0, 0, GGML_TYPE_F32, n_embed);
+    /*  aa */ ctx_size_add_tensor(ctx_size, 2, 1, GGML_TYPE_F32, n_embed);
+    /*  bb */ ctx_size_add_tensor(ctx_size, 1, 1, GGML_TYPE_F32, n_embed);
+    /*  pp */ ctx_size_add_tensor(ctx_size, 0, 0, GGML_TYPE_F32, n_embed);
+
+    /* wkv */ ctx_size_add_tensor(ctx_size, 1, 0, GGML_TYPE_F32, n_embed);
+    /*   x */ ctx_size_add_tensor(ctx_size, 2, 1, GGML_TYPE_F32, n_embed);
```

Comment:
I hope this will go away when ggml provides automatic way for calculating/allocating context...

Reply:
I think it's pretty but the amount of work was very high so yes I do hope that it can go away